src/jsifier.js

```diff
@@ -100,6 +109,68 @@ function JSify(data, functionsOnly) {
     return snippet;
   }
 
+  // Generates a function that invokes a proxied function call from the calling thread to the main browser thread.
+  function generateProxiedCallInvoker(sig, sync /*async if false*/) {
+    if (sig.length == 0) throw 'Function signature cannot be empty!';
+    function argsList(num) { // ", p0, p1, p2, p3, p4"
+      var s = '';
+      for(var i = 0; i < num; ++i) s += ', p' + i;
+      return s;
+    }
+
+    var func = "function _emscripten_" + (sync ? '' : 'a') + 'sync_run_in_browser_thread_' + sig + '(func' + argsList(sig.length-1) + ') {\n';
+    if (sync) func += '  var waitAddress = Runtime.stackSave();\n';
+
+    function sizeofType(t) {
+      switch(t) {
+        case 'd':
+//        case 'I': // int64  // TODO: For wasm, we'll have to have something like this
+          return 8;
+        case 'i':
+        case 'f':
+          return 4;
+        case 'v':
+          return 0;
+        // TODO: Smaller sizes?
+        default:
+          throw 'unsupported type in signature: ' + t;
+      }
+    }
+
+    var sizeofReturn = sizeofType(sig[0]);
+    if (sync) {
+      if (sizeofReturn == 4) func += '  var returnValue = waitAddress + 4;\n';
+      else if (sizeofReturn == 8) func += '  var returnValue = waitAddress + 8;\n'; // Retain alignment of each type
+    }
+
+    if (sync) func += '  Atomics.store(HEAP32, waitAddress >> 2, 0);\n';
+
+    function argsDict(num) { // ", p0: p0, p1: p1, p2: p2, p3: p3, p4: p4"
+      var s = '';
+      for(var i = 0; i < num; ++i) s += ', p' + i + ': p' + i;
+      return s;
+    }
+
+    // This is ad-hoc numbering scheme to map signatures to IDs, and must agree with call handler in src/library_pthread.js.
+    // Once proxied function calls no longer go through postMessage()s but instead in the heap, this will need to change, since int vs float will matter.
+    var functionCallOrdinal = sig.length + (sig[0] == 'd' ? 20 : 0);
```

Comment:
i still think the 20 is bad - couldn't that become invalid if we change memory layout? or am i missing something?

at minimum lets put a big comment with XXX FIXME on this.

Reply:
The number `20` is more like a magic ID, it only needs to match in this spot in `src/jsifier.js` and then in `src/pthread-main.js` (in https://github.com/kripken/emscripten/pull/5531/files#diff-db41bea94577c2dd9b0eef0308b06cf9R272). There is nothing about the memory layout itself that would interact with this. Only limit to number 20 is if there's a function that needs proxying that would take in more than 20 input parameters, then we'd have to bump up the number.

Basically this is turning all these signature strings `'vii', 'viii', 'iii', 'dii'` into ID numbers, so that the receiving code knows what kind of function signature it should be calling.

Reply:
I see, so it just cares about the # of params and whether the return value is float? maybe add a comment with that?

also, how about storing 20 in a const, and also asserting that the `sig.length` is shorter than it?

Reply:
Good suggestions, improved this. Also while reading, found a yet unimplemented case in handling there, fixed that, and also changed the number from 20 to 32 to be more programmer-friendly :)